onnxruntime/core/session/inference_session.h

```diff
@@ -733,6 +733,11 @@ class InferenceSession {
   // The life-cycle of the cache itself is maintained by the user and the user will ensure
   // the cache is valid until any session reliant on it is still in scope.
   PrepackedWeightsContainer* prepacked_weights_container_ = nullptr;
+
+  // Cache the CUDA EP instance if the user has configured the EP to capture a CUDA graph
+  // for the model and all the necessary criteria for CUDA graph capture has been met.
+  // At Run() time, if this member is not nullptr, simply invoke ReplayGraph() using it.
+  IExecutionProvider* cached_cuda_execution_provider_for_cuda_graph_replay_ = nullptr;
```

Comment:
Is there a better way to do this? This looks like something hardcoded/special cased just for cuda ep. Can there be a case where there are multiple EPs participating in the Run() and some of them (not just one) can be replayed in the subsequent runs? Say when tensorrt and cuda ep are both present. How would we support that? @hariharans29

Reply:
You mean how does a user pass-in a CUDA EP and a TensorRT EP instance to use in a session and turns on graph capturing for both and subsequently chooses one of the EPs to do the replay in Run() ? Currently, this is not supported. We need to think of a design for that.  We may have to use a RunOption for the user to choose which EP does the replaying. 

`This looks like something hardcoded/special cased just for cuda ep` : It is. We only set this if the entire graph has been partitioned to the CUDA EP and the user has requested usage of cuda graphs. Unfortunately, there isn't a good place/class for this variable to live in apart from `InferenceSession`, given that its core usage is in `InferenceSession::Run()`. The alternative is to remove this variable and add logic in Run() such that if Run() is invoked by the user, then check each EP in the priority order and check if a graph has been captured by the EP and if so, invoke `ReplayGraph()` on it. The downside is that it adds a runtime check in Run() that may not be as cheap as `if (cached_cuda_execution_provider_for_cuda_graph_replay_ )`. 

Open to other ideas though.

Reply:
For the sake of more clarity one way to do this would be: InferenceSession can store 2 members: an enum called run_mode (defaulted to NORMAL) and the EP enum to be replayed. This flag would be set to REPLAY along with storing the corresponding EP enum. During Run(), if the mode is REPLAY, you follow a different path: look up the EP corresponding to the EP enum and call Replay on it.

Reply:
`struct CachedExecutionProviderForGraphReplay` is added to avoid the hard code.